pkg/client/query/client_test.go

```diff
@@ -0,0 +1,439 @@
+		t.Run(testQuery(queryIdx), func(t *testing.T) {
+			var (
+				readObserverEventsTimeout = 300 * time.Millisecond
+				// readEventCounter is the  number of events which have been
+				// received from the connection since the subtest started.
+				readEventCounter int
+				// handleEventsLimit is the total number of events to send and
+				// receive through the query client's observable for this subtest.
+				handleEventsLimit = 1000
+				// handleEventsCounter is the number of events which have been
+				// received from the current query's observer channel since the
+				// (sub-)test started.
+				handleEventCounter int
+				// maxHandleEventTimes is the maximum number of times ReadEvent is
+				// expected to be called. It should be closed to handleMsgLimit but
+				// includes a +3% tolerance as it takes some time for concurrent context
+				// cancellation to propagate.
+				maxReadEventCount = handleEventsLimit * 105 / 100
+				queryLimit        = 1
+				connClosedMu      sync.Mutex
+				connClosed        bool
+				// done is used to signal when the subtest is complete
+				done = make(chan struct{}, 1)
+				// queryIdx is the index of the current query where each subtest
+				// exercises a unique query
+				queryIdx int
+			)
+
+			ctx, cancelQuery := context.WithCancel(ctx)
+			ctrl := gomock.NewController(t)
+			connMock := mockclient.NewMockConnection(ctrl)
+			// `Connection#Close()` should be called once for each subscription.
+			connMock.EXPECT().Close().
+				DoAndReturn(func() error {
+					connClosedMu.Lock()
+					defer connClosedMu.Unlock()
+
+					connClosed = true
+					return nil
+				}).
+				Times(1)
+			// `Connection#WriteJSON()` should be called once for each subscription.
+			connMock.EXPECT().WriteJSON(gomock.Any()).
+				Return(nil).
+				Times(1)
+			// `Connection#ReadEvent()` should be called once for each message plus
+			// one as it blocks in the loop which calls msgHandler after reading the
+			// last message.
+			connMock.EXPECT().ReadEvent().
+				DoAndReturn(func() ([]byte, error) {
+					connClosedMu.Lock()
+					defer connClosedMu.Unlock()
+
+					if connClosed {
+						return nil, errConnClosed
+					}
+
+					event := testEvent(readEventCounter)
+					readEventCounter++
+					return []byte(event), nil
+				}).
+				MaxTimes(maxReadEventCount).
+				MinTimes(handleEventsLimit)
+
+			dialerMock := mockclient.NewMockDialer(ctrl)
+			// `Dialer#DialContext()` should be called once for each subscription.
+			dialerMock.EXPECT().DialContext(gomock.Any(), gomock.Any()).
+				Return(connMock, nil).
+				Times(queryLimit)
+
+			// set up query client
+			dialerOpt := query.WithDialer(dialerMock)
+			queryClient := query.NewQueryClient("", dialerOpt)
+
+			// set up query observer
+			eventObservable, errCh := queryClient.EventsObservable(ctx, testQuery(queryIdx))
+			eventObserver := eventObservable.Subscribe(ctx)
+
+			// concurrently consume events from the observer channel
+			go func() {
```

Comment:
Consider refactoring/reusing [`goNotifiedOrTimedOutFactory`](https://github.com/pokt-network/poktroll/pull/31/files#diff-c8136d7c19962dcc168b21e249e0c612450ecc04bc778aa6bfd1946407bcce57R354) from #31.

Reply:
Alternatively, `goNotifiedOrTimedOutFactory` might be reducible once the `Either` primitive becomes available.